### PR TITLE
experiment: io_uring support

### PR DESCRIPTION
**Pull request by littledivy**
This PR does initial integration for using io_uring in Deno. See https://github.com/denoland/deno/issues/16232

Implements specialized tokio_iouring ops for `writeFile` and `readFile`. The tokio 
`current_thread` runtime is managed by `tokio_iouring::start`. We want to continue 
adding more ops specialized for iouring.

Additionally, all of io_uring functionality is behind the `enable_iouring` Cargo feature 
flag on `deno_runtime`. This is so that we do not distrupt workflow for people working 
on this repo. 

To compile Deno with io_uring support:
```
cargo build --release --features "deno_runtime/enable_iouring"
```

### `tokio_iouring`

It's early but has amazing integration with tokio's current_thread runtime. Blocked on the following tokio_iouring PRs:

- https://github.com/tokio-rs/tokio-uring/pull/44
- https://github.com/tokio-rs/tokio-uring/pull/131
- https://github.com/tokio-rs/tokio-uring/pull/132
- https://github.com/tokio-rs/tokio-uring/pull/133
- API for `opcode::Statx`

See [stuff that Deno needs](https://github.com/tokio-rs/tokio-uring/compare/master...littledivy:tokio-uring:deno) (ugly code alert).

Is it the right choice for Deno? This is what we aim to find out in the coming days with the experimental compilation flag.

### Benchmarks

4x improvement in `await Deno.writeFile(file, new Uint8Array(1024 * 1024 * 10))`:

```
divy@divy:~/deno$ target/release/deno run -A --unstable cli/bench/write_file.js # io_uring
time 433 ms rate 230
time 548 ms rate 182
time 540 ms rate 185
time 565 ms rate 176
^C

divy@divy:~/deno$ ~/.deno/bin/deno run -A --unstable cli/bench/write_file.js # main
time 2696 ms rate 37
time 2857 ms rate 35
time 2859 ms rate 34
time 2861 ms rate 34
^C
```

<details open><summary> main writeFile flamegraph </summary>

main_write_file

</details>
<details open><summary> io_uring writeFile flamegraph </summary>

io_uring_write_file

</details>

### CODE REVIEW

File: runtime/ops/asyncfs/iouring/mod.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  {
+    let path = Path::new(&path);
+    let mut state = state.borrow_mut();
+    state
+      .borrow_mut::<Permissions>()
+      .read
+      .check(path, Some("Deno.readFile()"))?;
+  }
+
+  // TODO(@littledivy): Stat file to get the size.
```
Comment (littledivy) [inline runtime/ops/asyncfs/iouring/mod.rs:131]: this is WIP lol. `resize` below may explode allocations for large files. We should stat the file for size.

File: serde_v8/Cargo.toml [hunk 2]
```diff
@@ -12,12 +12,18 @@ description = "Rust to V8 serialization and deserialization"
 [lib]
 path = "lib.rs"
 
+[features]
+default = []
+# Requires a recent Linux kernel version (atleast )
```
Comment (littledivy) [inline serde_v8/Cargo.toml:17]:
```suggestion
# Requires a recent Linux kernel version (atleast 5.6; tested on >5.11.0)
```

I think 5.6 is safe to assume? Read op was added in 5.6. I'm on 5.15.0 and tokio_iouring seems to suggest >5.11.0

File: runtime/ops/asyncfs/iouring/mod.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  let fut = async move {
+    let path = Path::new(&path);
+    let file = File::open(&path).await?;
+    let mut buf = vec![0; 1024 * 16];
+    let mut offset: u64 = 0;
+    loop {
+      let (res, mut ret) =
+        file.read_at(buf.split_off(offset as usize), offset).await;
```
Comment (bnoordhuis) [inline runtime/ops/asyncfs/iouring/mod.rs:139]: `buf.split_off()` is wildly inefficient for the `offset > 0` case (creates new vec, copies data, leaves old vec's capacity unchanged, i.e., wasted.)

Perhaps a better way to handle this is to maintain a vec of vecs, then join them at the end. That's `O(n)` but that's better than `O(n*m)`.

File: serde_v8/Cargo.toml (hunk 2 above)
Comment (bnoordhuis) [reply serde_v8/Cargo.toml:17]: IORING_OP_READV and IORING_OP_READ_FIXED - what maps to the Readv and ProvideBuffers ops, if I read the the io_uring crate's source code correctly - were added in 5.1.

**Review by ry — COMMENTED** [review]
This patch is cleanly implemented, but ties us ever more to Tokio. This work might need to be undone if we went in a Crimson or pollio direction.

File: runtime/ops/asyncfs/iouring/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (littledivy) [reply runtime/ops/asyncfs/iouring/mod.rs:131]: Stat size is simply a preallocation hint. We should stop reading only on EOF

**Comment by espoal** [comment]
question: if I understood correctly [here](https://github.com/denoland/deno/blob/212b7dd6da487c070229b6348ec7907b4fecbcf9/runtime/ops/fs.rs#L140) you use the unix syscall to open a file. Why didn't you use `IORING_OP_OPENAT`? that would save a syscall.

**Comment by littledivy** [comment]
@espoal that is a syncronous op for `Deno.openSync`.

**Comment by littledivy** [comment]
we are trying to optimize async I/O here. See [`runtime/ops/asyncfs/iouring/mod.rs`](https://github.com/littledivy/deno/blob/io_uring/runtime/ops/asyncfs/iouring/mod.rs)

**Comment by lino-levan** [comment]
@littledivy are there any updates? All of the PRs you mentioned blocking this change are merged now. This would be big for linux perf.

**Comment by KnorpelSenf** [comment]
> This patch is cleanly implemented, but ties us ever more to Tokio. This work might need to be undone if we went in a Crimson or pollio direction.

Is this the reason why this was abandoned?

**Comment by aapoalas** [comment]
@littledivy @ry Do we expect this PR to be taken up again at some point, or will it be abandoned altogether? ie. Should we close this?

**Comment by littledivy** [comment]
I still want to do this but I don't have time to prioritize this atm. PRs welcome

**Comment by rektide** [comment]
What are some take that would help this pr?

Each do the three blocking tokio-uring PR are merged now. Presumably updating this branch to use those would help?

**Comment by bnoordhuis** [comment]
As a data point: after adding io_uring support to libuv we've been receiving a veritable fountain of reports from people hitting kernel bugs, from 100% CPU consumption to file operations not working on CIFS mounts.

Our kernel blacklist is long and specific, and still growing longer and more specific; see e.g. https://github.com/libuv/libuv/pull/4224 from just yesterday. I honestly can't really recommend io_uring right now.

**Comment by espoal** [comment]
@bnoordhuis `io_uring` is not thought to handle network mounted file systems. 
I'm surprised of your report, because I built a cloud scale database on top of `io_uring` and I don't see those problems. Could you provide more sources? I will take time personally to debug and eventually propose a patch for the interface.

The ticket you mention do not refer to a bug, but to the fact that the interface of `io_uring` in kernel 6.1 improved dramatically, both in term of ergonomics and performance, hence why the community try to aim for that version at least.

**Comment by bnoordhuis** [comment]
@espoal search the nodejs/node and libuv/libuv repos issue trackers, you'll find plenty of issues, some still open

> I built a cloud scale database on top of io_uring and I don't see those problems

That probably means you're either targeting only really new kernels, or your user base is a fraction of node and libuv. Both projects have _huge_ user bases and people hit (and report) the weirdest bugs. The war stories I could tell...

**Comment by espoal** [comment]
@bnoordhuis that's a very fair point. I also noticed a huge stability improvement from 6.0 to 6.1

Sorry if I came out as rude, but I felt that you were attacking my single favourite kernel interface :smile: thanks for taking the time to elaborate :heart:

**Comment by bnoordhuis** [comment]
No worries, didn't come across as rude.

**Pull request closed without merging**